File: examples/rust/common/src/rpc.rs
use reqwest::blocking::Client;
use serde_json::{json, Value};
use std::env;
use std::path::{Path, PathBuf};

pub struct SandboxConfig {
    pub rpc_url: String,
    pub rpc_user: String,
    pub rpc_password: String,
    pub rpc_wallet: String,
    pub zmq_rawblock: String,
    pub zmq_rawtx: String,
}

fn find_repo_root() -> PathBuf {
    let mut current = env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    loop {
        if current.join(".git").exists() || current.join(".env.example").exists() {
            return current;
        }
        if !current.pop() {
            break;
        }
    }
    PathBuf::from(".")
}

pub fn get_config() -> SandboxConfig {
    let repo_root = find_repo_root();
    let env_file = repo_root.join(".env");
    let env_example = repo_root.join(".env.example");

    if env_file.exists() {
        let _ = dotenvy::from_path(&env_file);
    } else if env_example.exists() {
        let _ = dotenvy::from_path(&env_example);
    }

    SandboxConfig {
        rpc_url: env::var("BITCOIN_RPC_URL").unwrap_or_else(|_| "http://127.0.0.1:18443".to_string()),
        rpc_user: env::var("BITCOIN_RPC_USER").unwrap_or_else(|_| "bitcoinrpc".to_string()),
        rpc_password: env::var("BITCOIN_RPC_PASSWORD").unwrap_or_else(|_| "bitcoinrpcpassword".to_string()),
        rpc_wallet: env::var("BITCOIN_RPC_WALLET").unwrap_or_else(|_| "lab".to_string()),
        zmq_rawblock: env::var("BITCOIN_ZMQ_RAWBLOCK").unwrap_or_else(|_| "tcp://127.0.0.1:28332".to_string()),
        zmq_rawtx: env::var("BITCOIN_ZMQ_RAWTX").unwrap_or_else(|_| "tcp://127.0.0.1:28333".to_string()),
    }
}

#[derive(Clone)]
pub struct BitcoinRPC {
    pub base_url: String,
    pub user: String,
    pub password: String,
    pub wallet: Option<String>,
    client: Client,
}

impl BitcoinRPC {
    pub fn new() -> Self {
        let config = get_config();
        Self {
            base_url: config.rpc_url.trim_end_matches('/').to_string(),
            user: config.rpc_user,
            password: config.rpc_password,
            wallet: None,
            client: Client::new(),
        }
    }

    pub fn with_wallet(wallet: &str) -> Self {
        let mut rpc = Self::new();
        rpc.wallet = Some(wallet.to_string());
        rpc
    }

    pub fn url(&self) -> String {
        if let Some(ref w) = self.wallet {
            format!("{}/wallet/{}", self.base_url, w)
        } else {
            self.base_url.clone()
        }
    }

    pub fn for_wallet(&self, wallet: &str) -> Self {
        let mut cloned = self.clone();
        cloned.wallet = Some(wallet.to_string());
        cloned
    }

    pub fn call(&self, method: &str, params: Vec<Value>) -> Result<Value, String> {
        let payload = json!({
            "jsonrpc": "1.0",
            "id": "bitcoin-sandbox-rs",
            "method": method,
            "params": params
        });

        let resp = self
            .client
            .post(self.url())
            .basic_auth(&self.user, Some(&self.password))
            .json(&payload)
            .send()
            .map_err(|e| format!("Could not connect to Bitcoin node at {}: {}", self.url(), e))?;

        let body: Value = resp
            .json()
            .map_err(|e| format!("Failed to parse response JSON: {}", e))?;

        if let Some(err) = body.get("error").filter(|e| !e.is_null()) {
            return Err(format!("Bitcoin RPC Error: {}", err));
        }

        Ok(body.get("result").cloned().unwrap_or(Value::Null))
    }

    pub fn get_rest(&self, endpoint: &str) -> Result<Value, String> {
        let clean = endpoint.trim_start_matches('/');
        let url = format!("{}/rest/{}", self.base_url, clean);

        let resp = self
            .client
            .get(&url)
            .send()
            .map_err(|e| format!("Could not connect to REST endpoint at {}: {}", url, e))?;

        if !resp.status().is_success() {
            return Err(format!("REST endpoint returned status: {}", resp.status()));
        }

        resp.json().map_err(|e| format!("Failed to parse REST JSON: {}", e))
    }
}

File: examples/go/common/rpc.go
package common

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/joho/godotenv"
)

type Config struct {
	RPCURL      string
	RPCUser     string
	RPCPassword string
	RPCWallet   string
	ZMQRawBlock string
	ZMQRawTx    string
}

func findRepoRoot() string {
	dir, err := os.Getwd()
	if err != nil {
		return "."
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, ".git")); err == nil {
			return dir
		}
		if _, err := os.Stat(filepath.Join(dir, ".env.example")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return "."
}

func GetConfig() Config {
	repoRoot := findRepoRoot()
	envPath := filepath.Join(repoRoot, ".env")
	envExamplePath := filepath.Join(repoRoot, ".env.example")

	if _, err := os.Stat(envPath); err == nil {
		_ = godotenv.Load(envPath)
	} else if _, err := os.Stat(envExamplePath); err == nil {
		_ = godotenv.Load(envExamplePath)
	}

	getEnv := func(key, defaultVal string) string {
		if v := os.Getenv(key); v != "" {
			return v
		}
		return defaultVal
	}

	return Config{
		RPCURL:      getEnv("BITCOIN_RPC_URL", "http://127.0.0.1:18443"),
		RPCUser:     getEnv("BITCOIN_RPC_USER", "bitcoinrpc"),
		RPCPassword: getEnv("BITCOIN_RPC_PASSWORD", "bitcoinrpcpassword"),
		RPCWallet:   getEnv("BITCOIN_RPC_WALLET", "lab"),
		ZMQRawBlock: getEnv("BITCOIN_ZMQ_RAWBLOCK", "tcp://127.0.0.1:28332"),
		ZMQRawTx:    getEnv("BITCOIN_ZMQ_RAWTX", "tcp://127.0.0.1:28333"),
	}
}

type BitcoinRPC struct {
	BaseURL  string
	User     string
	Password string
	Wallet   string
	client   *http.Client
}

func NewBitcoinRPC() *BitcoinRPC {
	cfg := GetConfig()
	return &BitcoinRPC{
		BaseURL:  strings.TrimRight(cfg.RPCURL, "/"),
		User:     cfg.RPCUser,
		Password: cfg.RPCPassword,
		client:   &http.Client{},
	}
}

func (r *BitcoinRPC) ForWallet(walletName string) *BitcoinRPC {
	return &BitcoinRPC{
		BaseURL:  r.BaseURL,
		User:     r.User,
		Password: r.Password,
		Wallet:   walletName,
		client:   r.client,
	}
}

func (r *BitcoinRPC) URL() string {
	if r.Wallet != "" {
		return fmt.Sprintf("%s/wallet/%s", r.BaseURL, r.Wallet)
	}
	return r.BaseURL
}

type rpcRequest struct {
	JSONRPC string        `json:"jsonrpc"`
	ID      string        `json:"id"`
	Method  string        `json:"method"`
	Params  []interface{} `json:"params"`
}

type rpcResponse struct {
	Result json.RawMessage `json:"result"`
	Error  *struct {
		Code    int    `json:"code"`
		Message string `json:"message"`
	} `json:"error"`
	ID string `json:"id"`
}

func (r *BitcoinRPC) Call(method string, params ...interface{}) (json.RawMessage, error) {
	if params == nil {
		params = []interface{}{}
	}
	reqBody, err := json.Marshal(rpcRequest{
		JSONRPC: "1.0",
		ID:      "bitcoin-sandbox-go",
		Method:  method,
		Params:  params,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to marshal JSON-RPC request: %w", err)
	}

	req, err := http.NewRequest("POST", r.URL(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.SetBasicAuth(r.User, r.Password)

	resp, err := r.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("could not connect to Bitcoin node at %s: %w", r.URL(), err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	var rpcResp rpcResponse
	if err := json.Unmarshal(body, &rpcResp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON-RPC response: %w", err)
	}

	if rpcResp.Error != nil {
		return nil, fmt.Errorf("bitcoin RPC Error (%d): %s", rpcResp.Error.Code, rpcResp.Error.Message)
	}

	return rpcResp.Result, nil
}

func (r *BitcoinRPC) GetREST(endpoint string) (json.RawMessage, error) {
	clean := strings.TrimLeft(endpoint, "/")
	url := fmt.Sprintf("%s/rest/%s", r.BaseURL, clean)

	resp, err := r.client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("could not connect to REST endpoint at %s: %w", url, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("REST request failed with status: %d", resp.StatusCode)
	}

	return io.ReadAll(resp.Body)
}
